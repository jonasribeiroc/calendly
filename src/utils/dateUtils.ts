import {
    addMonths,
    subMonths,
    parseISO,
    isSameDay,
    format,
    parse,
    isValid,
    getYear as getYearFns,
    getMonth as getMonthFns,
    getDay,
    lastDayOfMonth,
    isSameMonth as isSameMonthFns,
    startOfMonth,
    endOfMonth,
    getDate,
} from 'date-fns';

function getDateUTC(date?: Date): Date {
    const baseDate = date instanceof Date ? date : new Date();

    return new Date(Date.UTC(
        baseDate.getUTCFullYear(),
        baseDate.getUTCMonth(),
        baseDate.getUTCDate(),
        baseDate.getUTCHours(),
        baseDate.getUTCMinutes(),
        baseDate.getUTCSeconds(),
        baseDate.getUTCMilliseconds()
    ));
}

function isSameMonth(date1: Date, date2: Date): boolean {
    return isSameMonthFns(date1, date2);
}

function isValidDateString(str: string, formats: string[]): boolean {
    return formats.some(formatStr => {
        const parsedDate = parse(str, formatStr, getDateUTC());

        return isValid(parsedDate) && format(parsedDate, formatStr) === str;
    });
}

function getDateFromString(str: string | null, dateFormat: string): Date | undefined {
    if (str && isValidDateString(str, [dateFormat])) {
        return getDateUTC(new Date(str.replace('-', ',')));
    }
}

function getDateString(date: Date, formatStr?: string): string {
    return format(date, formatStr || 'yyyy-MM');
}

function getYear(date: Date): number {
    return getYearFns(date);
}

function getMonth(date: Date): number {
    return getMonthFns(date);
}

function getMonthName(date: Date): string {
    return format(date, 'MMMM');
}

function getMonthDays(date: Date): number {
    return getDate(lastDayOfMonth(date));
}

function getWeekDay(date: Date): number {
    return getDay(date);
}

function getWeekDayName(date: Date): string {
    return format(date, 'EEEE');
}

function getMonthDay(date: Date): number {
    return Number(format(date, 'd'));
}

function getTime(date: Date): string {
    return format(date, 'HH:mm');
}

function getFirstDayOfMonth(date: Date): Date {
    const firstDay = startOfMonth(date);

    return getDateUTC(new Date(getYear(firstDay), getMonth(firstDay), getMonthDay(firstDay), 0, 0, 0, 0));
}

function getLastDayOfMonth(date: Date): Date {
    const lastDay = endOfMonth(date);

    return getDateUTC(new Date(getYear(lastDay), getMonth(lastDay), getMonthDay(lastDay), 23, 59, 59, 999));
}

function getAvailableDays(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameMonth(d, date))
}

function getAvailableTimes(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameDay(d, date))
}

function getNextMonth(date: Date): Date {
    return addMonths(date, 1);
}

function getPreviousMonth(date: Date): Date {
    return subMonths(date, 1);
}

export {
    isSameMonth,
    isValidDateString,
    getDateUTC,
    getDateFromString,
    getDateString,
    getYear,
    getMonth,
    getMonthName,
    getMonthDays,
    getWeekDay,
    getWeekDayName,
    getMonthDay,
    getTime,
    getFirstDayOfMonth,
    getLastDayOfMonth,
    getAvailableDays,
    getAvailableTimes,
    getNextMonth,
    getPreviousMonth,
}