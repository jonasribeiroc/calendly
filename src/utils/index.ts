import {
    addMonths,
    subMonths,
    parseISO,
    isSameDay,
    format,
    parse,
    isValid,
    getYear as getYearFns,
    getMonth as getMonthFns,
} from 'date-fns';

function isSameYearMonth(str1?: string | null, str2?: string | null): boolean {
    if (!str1 || !str2) return false;

    const [year1, month1] = str1.split('-');
    const [year2, month2] = str2.split('-');

    return year1 === year2 && month1 === month2;
}

function isValidDateStr(str: string | null, formats: string[]): boolean {
    return formats.some(formatStr => {
        const parsedDate = parse(str || '', formatStr, new Date());

        return isValid(parsedDate) && format(parsedDate, formatStr) === str;
    });
}

function getDate(str?: string | null, formatStr?: 'yyyy-MM' | 'yyyy-MM-dd'): Date {
    return !!str && isValidDateStr(str, ['yyyy-MM-dd', 'yyyy-MM']) ? parse(str, formatStr || 'yyyy-MM', new Date()) : new Date();
}

function getDateString(date?: Date | null, formatStr?: 'yyyy-MM' | 'yyyy-MM-dd' | 'yyyy-MM-dd\'T\'HH:mm:ss' | "yyyy-MM-dd'T'HH:mm:ssxxx"): string {
    return format(date || new Date(), formatStr || 'yyyy-MM');
}

function getYear(date: Date): number {
    return getYearFns(date);
}

function getMonth(date: Date): number {
    return getMonthFns(date);
}

function getMonthName(date: Date): string {
    return format(date, 'MMMM');
}

function getMonthDays(date: Date): number {
    const year = getYear(date);
    const month = getMonth(date);

    return new Date(year, month, 0).getDate();
}

function getWeekDay(date: Date): number {
    return date.getUTCDay();
}

function getWeekDayName(date: Date): string {
    return date.toLocaleString('en-US', { weekday: 'long' });
}

function getDay(date: Date): number  {
    return Number(date.toLocaleString('en-US', { day: 'numeric' }));
}

function getTime(date: Date): string {
    return format(date, 'hh:mm');
}

function getFirstDayOfMonth(date: Date): Date {
    return new Date(date.getFullYear(), date.getMonth(), 1);
  }

function getAvailableDays(availableDateTimes: string[], date: Date): number[] {
    const daysSet = new Set<number>();
    const year = getYear(date);
    const month = getMonth(date);

    availableDateTimes.forEach((dateTime) => {
        const date = new Date(dateTime);
        const dateYear = getYear(date);
        const dateMonth = getMonth(date);

        if (dateYear === year && dateMonth === month) {
            daysSet.add(getDay(date));
        }
    });

    return Array.from(daysSet).sort((a, b) => a - b);
}

function getAvailableTimes(availableDateTimes: string[], date: Date): Date[] {
    return availableDateTimes
        .map(dateTime => parseISO(dateTime))
        .filter(d => isSameDay(d, date))
}

function getNextMonth(date: Date): Date {
    return addMonths(date, 1);
}

function getPreviousMonth(date: Date): Date {
    return subMonths(date, 1);
}

export {
    isSameYearMonth,
    isValidDateStr,
    getDate,
    getDateString,
    getYear,
    getMonth,
    getMonthName,
    getMonthDays,
    getWeekDay,
    getWeekDayName,
    getDay,
    getTime,
    getFirstDayOfMonth,
    getAvailableDays,
    getAvailableTimes,
    getNextMonth,
    getPreviousMonth,
}