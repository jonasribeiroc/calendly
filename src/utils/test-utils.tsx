import React, { ReactNode } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../theme';

interface AllTheProvidersProps {
  children: ReactNode;
}

const AllTheProviders = ({ children }: AllTheProvidersProps) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
