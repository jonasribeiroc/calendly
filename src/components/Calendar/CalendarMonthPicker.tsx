import React from 'react';
import { FiChevronLeft, FiChevronRight } from 'react-icons/fi';
import styled from 'styled-components';
import { getMonthName, getNextMonth, getPreviousMonth, getYear } from '../../utils/dateUtils';

const CalendarMonthPickerContainer = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
    font-weight: 500;
    margin-bottom: 20px;
    position: relative;
`;

const CalendarMonthPickerButton = styled.button`
    position: absolute;
    border: none;
    background: none;
    cursor: pointer;
    font-size: 18px;
    color: #666;

    &:hover {
      color: #333;
    }
`;

const CalendarMonthPickerPrevButton = styled(CalendarMonthPickerButton)`
    left: 0;
`;

const CalendarMonthPickerNextButton = styled(CalendarMonthPickerButton)`
    right: 0;
`;

const CalendarMonthPickerTitle = styled.span`
    font-size: 15px;
`;

interface CalendarMonthPickerProps {
    month: Date,
    onChange: (date: Date) => void
}

export const CalendarMonthPicker: React.FC<CalendarMonthPickerProps> = ({
    month,
    onChange
}) => {

    const handlePrev = () => {
        const newDate = getPreviousMonth(month);
        onChange(newDate);
    };

    const handleNext = () => {
        const newDate = getNextMonth(month);
        onChange(newDate);
    };

    return (
        <CalendarMonthPickerContainer>
            <CalendarMonthPickerPrevButton aria-label='previous month' onClick={handlePrev}>
                <FiChevronLeft />
            </CalendarMonthPickerPrevButton>
            <CalendarMonthPickerTitle>{getMonthName(month)} {getYear(month)}</CalendarMonthPickerTitle>
            <CalendarMonthPickerNextButton aria-label='next month' onClick={handleNext}>
                <FiChevronRight />
            </CalendarMonthPickerNextButton>
        </CalendarMonthPickerContainer>
    );
}

