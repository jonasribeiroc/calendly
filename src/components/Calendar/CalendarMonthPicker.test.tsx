// CalendarMonthPicker.test.tsx

import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { CalendarMonthPicker } from './CalendarMonthPicker';
import * as dateUtils from '../../utils/dateUtils';
import { FiChevronLeft, FiChevronRight } from 'react-icons/fi';

// Mock das funções de dateUtils
jest.mock('../../utils/dateUtils');

// Mock dos ícones do react-icons
jest.mock('react-icons/fi', () => ({
  FiChevronLeft: () => <div data-testid="fi-chevron-left" />,
  FiChevronRight: () => <div data-testid="fi-chevron-right" />,
}));

describe('CalendarMonthPicker Component', () => {
  // Tipagem dos mocks para TypeScript
  const mockedDateUtils = dateUtils as jest.Mocked<typeof dateUtils>;

  // Função mock para onChange
  const mockOnChange = jest.fn();

  // Dados de teste
  const mockMonth = new Date('2025-04-15T00:00:00Z'); // Abril de 2025

  beforeEach(() => {
    jest.clearAllMocks();

    // Mock das funções dateUtils
    mockedDateUtils.getMonthName.mockImplementation((date: Date) => {
      return date.toLocaleString('default', { month: 'long' });
    });

    mockedDateUtils.getYear.mockImplementation((date: Date) => {
      return date.getUTCFullYear();
    });

    mockedDateUtils.getNextMonth.mockImplementation((date: Date) => {
      const year = date.getUTCFullYear();
      const month = date.getUTCMonth();
      return new Date(Date.UTC(month === 11 ? year + 1 : year, (month + 1) % 12, 1));
    });

    mockedDateUtils.getPreviousMonth.mockImplementation((date: Date) => {
      const year = date.getUTCFullYear();
      const month = date.getUTCMonth();
      return new Date(Date.UTC(month === 0 ? year - 1 : year, (month - 1 + 12) % 12, 1));
    });
  });

  it('renders the correct month and year', () => {
    render(<CalendarMonthPicker month={mockMonth} onChange={mockOnChange} />);

    const monthYearTitle = screen.getByText(`${mockedDateUtils.getMonthName(mockMonth)} ${mockedDateUtils.getYear(mockMonth)}`);
    expect(monthYearTitle).toBeInTheDocument();
  });

  it('renders the previous and next buttons', () => {
    render(<CalendarMonthPicker month={mockMonth} onChange={mockOnChange} />);

    const prevButton = screen.getByRole('button', { name: /previous month/i });
    const nextButton = screen.getByRole('button', { name: /next month/i });

    expect(prevButton).toBeInTheDocument();
    expect(nextButton).toBeInTheDocument();

    // Verificar se os ícones estão presentes
    expect(screen.getByTestId('fi-chevron-left')).toBeInTheDocument();
    expect(screen.getByTestId('fi-chevron-right')).toBeInTheDocument();
  });

  it('calls onChange with the previous month when prev button is clicked', () => {
    render(<CalendarMonthPicker month={mockMonth} onChange={mockOnChange} />);

    const prevButton = screen.getByRole('button', { name: /previous month/i });

    fireEvent.click(prevButton);

    const expectedPreviousMonth = mockedDateUtils.getPreviousMonth(mockMonth);
    expect(mockOnChange).toHaveBeenCalledWith(expectedPreviousMonth);
  });

  it('calls onChange with the next month when next button is clicked', () => {
    render(<CalendarMonthPicker month={mockMonth} onChange={mockOnChange} />);

    const nextButton = screen.getByRole('button', { name: /next month/i });

    fireEvent.click(nextButton);

    const expectedNextMonth = mockedDateUtils.getNextMonth(mockMonth);
    expect(mockOnChange).toHaveBeenCalledWith(expectedNextMonth);
  });

  it('handles year boundaries correctly when navigating from December to January', () => {
    const december = new Date('2025-12-15T00:00:00Z');
    mockedDateUtils.getMonthName.mockReturnValue('December');
    mockedDateUtils.getYear.mockReturnValue(december.getUTCFullYear());

    render(<CalendarMonthPicker month={december} onChange={mockOnChange} />);

    const nextButton = screen.getByRole('button', { name: /next month/i });
    fireEvent.click(nextButton);

    const expectedNextMonth = new Date(Date.UTC(2026, 0, 1)); // Janeiro de 2026
    expect(mockOnChange).toHaveBeenCalledWith(expectedNextMonth);
  });

  it('handles year boundaries correctly when navigating from January to December', () => {
    const january = new Date('2025-01-15T00:00:00Z');
    mockedDateUtils.getMonthName.mockReturnValue('January');
    mockedDateUtils.getYear.mockReturnValue(january.getUTCFullYear());

    render(<CalendarMonthPicker month={january} onChange={mockOnChange} />);

    const prevButton = screen.getByRole('button', { name: /previous month/i });
    fireEvent.click(prevButton);

    const expectedPreviousMonth = new Date(Date.UTC(2024, 11, 1)); // Dezembro de 2024
    expect(mockOnChange).toHaveBeenCalledWith(expectedPreviousMonth);
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<CalendarMonthPicker month={mockMonth} onChange={mockOnChange} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
