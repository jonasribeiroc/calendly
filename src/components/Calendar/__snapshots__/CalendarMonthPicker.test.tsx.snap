// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarMonthPicker Component matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-blHHSb fJKPUX"
  >
    <button
      aria-label="previous month"
      class="sc-gtLWhw sc-egkSDF lorpGW fBSbpZ"
    >
      <div
        data-testid="fi-chevron-left"
      />
    </button>
    <span
      class="sc-dntaoT euESij"
    >
      April 2025
    </span>
    <button
      aria-label="next month"
      class="sc-gtLWhw sc-fAUdSK lorpGW gofhhg"
    >
      <div
        data-testid="fi-chevron-right"
      />
    </button>
  </div>
</DocumentFragment>
`;
